[WIP] Changes to allow optimization of mixing hard and soft logic

georgkrylov opened the pull request:
#### Description
These changes address problem of ODIN's inability to use smaller number of multipliers than it is available. The changes are designed to work with multipliers first, should be easy to extend to adders and in theory possible to extend too other blocks


#### Motivation and Context
This change is required since sometimes, the synthesis flow for fixed layout would break even if the logic could be attempted to be implemented in soft logic that is not used otherwise. Moreover, setting a limit to number of hard blocks for auto layouts might change  the shape of a device and could prevent from growing the device unnecessary.

#### How Has This Been Tested?
This is a very rough work in progress PR,  that does not pretend to be finalized. Thorough testing is still an open question. The CI seems to be unaffected by the introduced changes if the optimizations are not enabled, but the tests if it actually works or not are yet to be finalized

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

jeanlego reviewed (commented):
you seem to have added unnecessary files to this commit, we can't change the gitignore but just stick to the naming standard for test and it should fix the wrongly commited files (1345 files!!!) (0% directory??)

georgkrylov commented:
@jeanlego Right, I was demonstrating the results for the course. I will move the results to other repository.

jeanlego commented:
> @jeanlego Right, I was demonstrating the results for the course. I will move the results to other repository.

gotcha :) didn't mean to pressure you, just ping me when it's ready for review because I can't right now due to the too large PR (takes an eternity to load in the browser)

georgkrylov commented:
Hello @jeanlego, would you prefer having the test config files in this Pull Request or we should leave them as a subject for a separate PR?

jeanlego commented:
The test config should be in a separate PR

georgkrylov commented:
Hi @jeanlego , how does this shape look in terms of PR readability? Shall the task changes to the .pl script be a subject of a separate PR?

jeanlego commented:
This is already much better! 

If you think they can be split then yeah I would split them, just so I can review it faster . Thank you! The difference is noticable! Good job

jeanlego commented on ODIN_II/SRC/include/netlist_analyzer.h:
```diff
@@ -0,0 +1,34 @@
+#define MULT_OPTIMIZATION_WEIGHT_
+// Returns a random number
+int calculate_multiplier_aware_critical_path(nnode_t* node,netlist_t* netlist);
+#endif
```
try to leave an extra line at the end of a file. clang has a tendency to warn about this.

georgkrylov replied:
Probably, I am not using the pull request correctly. This file does not look like that in the latest commit... Anyway, an extra line will be added and shipped the next commit.

georgkrylov commented:
Hello, @jeanlego , I have attempted to make the PR look nicer. Other than writing more comments, I do not have any other ideas how to help merging process, could you please review the code and suggest possible improvements?

jeanlego commented on ODIN_II/SRC/partial_map.cpp:
```diff
@@ -186,13 +186,22 @@ void partial_map_node(nnode_t *node, short traverse_number, netlist_t *netlist)
-					instantiate_hard_adder(node, traverse_number, netlist);
-			}else{
-				instantiate_add_w_carry(node, traverse_number, netlist);
+			if (true == mixer->mixHardBlocksOfType(HardBlocksOptimizationTypesEnum::MULTIPLIERS) ||
```
Don't check true on a boolean make it your variable and function call explicit about them returning a true.

Don't use camel case, the only place camel case is used is in the parse AST, try to match the style of the file

Is specifying the enum necessary? I feel it adds complexity by adding more strict and enums to track. Could you use the globally declared operation_list?

georgkrylov replied:
I will present better phrasing in my next commits, tell me if you agree with changes

jeanlego commented on ODIN_II/SRC/partial_map.cpp:
```diff
@@ -186,13 +186,22 @@ void partial_map_node(nnode_t *node, short traverse_number, netlist_t *netlist)
+				}
+				break;
 			}
-			break;
```
The first if statement is gonna fall through without this break, avoid fallthrough with the break at the end of your case or put the non breaking logic here at the end and explicitly fallthrough by putting at least a //fallthrough comment

georgkrylov replied:
I "think" there's no fall through here because the true part of the statement is terminated with break, so is the else part. Alternatively, for readability purposes I can introduce a {} around if, but it seemed unreasonable to introduce {} for a single if statement.

jeanlego replied:
We always use the bracket wether it be for one line or Multiple. Just the way it was written to begin with, some files seem to stray away from that but that's because no on was enforcing the standard.

jeanlego replied:
To stick with the rest of the file it should look something like this
```c

		case ADD:
			if (mixer->can_be_hardened(MULTIPLY) || mixer->can_be_hardened(ADD) )
			{
				mixer->flag_for_hardenning(node,ADD);
			}
                        else
			{
				if (hard_adders && node->bit_width >= min_threshold_adder)
                                {
					instantiate_hard_adder(node, traverse_number, netlist);
				}
                                else
                                {
					instantiate_add_w_carry(node, traverse_number, netlist);
				}
			}
                        break;
```

georgkrylov replied:
Thank you, fixed! I will ship the changes in the next commit

jeanlego commented on ODIN_II/SRC/include/HardSoftLogicMixer.hpp:
```diff
@@ -0,0 +1,105 @@
+    // This array is composed of vectors, that store nodes that
+    // are potential candidates for performing mixing optimization
+    std::vector<nnode_t*>potentialHardBlockNodes[HardBlocksOptimizationTypesEnum::Count];
+    int _hardBlocksCount [HardBlocksOptimizationTypesEnum::Count];
```
Why the _ before the name? Unless it is to match the naming style of the file (they are mostly use to hide function behind define wrapper in odin_error.h) this makes it confusing

georgkrylov replied:
It is usually used as a prefix for private variable names as a visual indication that the scope of the variable is private to the class in which it is declared. (from stackoverflow). Shall I change it?

jeanlego commented on ODIN_II/SRC/HardSoftLogicMixer.cpp:
```diff
@@ -0,0 +1,256 @@
+	}		
+	for (int i = 0 ; i < _arch.grid_layouts.size();i++)
+	{
+		if (i == 1){
```
Keep your bracket on a separate line than the condition 
```C
If(a==b)
{
 Body
}
```

georgkrylov replied:
This is solved by applying clang formatter!

jeanlego commented on ODIN_II/SRC/HardSoftLogicMixer.cpp:
```diff
@@ -0,0 +1,256 @@
+	for (int i = 0 ; i < _arch.grid_layouts.size();i++)
+	{
+		if (i == 1){
+			std::cerr<<"Optimization for multiple layouts was never tested, from now on you are on your own"<<std::endl;
```
We have our own error reporting mechanism, error_message(...) Use that rather

jeanlego commented on ODIN_II/SRC/OdinGridAnalyzer.cpp:
```diff
@@ -0,0 +1,334 @@
+		std::cout<<"Start_expr\t" << grid_loc_def->x.start_expr<<"\t"<<startx
+				<<"\t"<<grid_loc_def->y.start_expr<<"\t"<<starty<<std::endl;
+
+		std::cout<<"End_expr\t" << grid_loc_def->x.end_expr<<"\t"<<endx
```
We use printf not cout throughout

jeanlego commented on ODIN_II/SRC/OdinGridAnalyzer.cpp:
```diff
@@ -0,0 +1,334 @@
+        //Check start against the device dimensions
+        // Start locations outside the device will never create block instances
+        if (startx > grid_width - 1) {
+			return;
```
Return at the end of the function.
Function are laid out 
```
Initialize variable
Body
Return
```

jeanlego commented on ODIN_II/SRC/include/odin_types.h:
```diff
@@ -532,6 +538,13 @@ struct nnet_t
+// the last element. The order of the elements 
+// in this enum also determines order of optimizations
+// ADDERS SHOULD BE AFTER MULTS
+enum HardBlocksOptimizationTypesEnum { MULTIPLIERS, ADDERS, Count};
```
Enums should be on multiple line (one value/line see types.h)

jeanlego commented on ODIN_II/SRC/netlist_analyzer.cpp:
```diff
@@ -0,0 +1,67 @@
+
+#include "netlist_analyzer.h"
+
+int calculate_multiplier_aware_critical_path(nnode_t* node,netlist_t* netlist){
```
Avoid nesting more than 4 bracket deep, if it's more than that, use functions with good descriptive name

jeanlego commented on ODIN_II/SRC/HardSoftLogicMixer.cpp:
```diff
@@ -0,0 +1,256 @@
+#include <cstdint>
+#include <cstring>
+#include "OdinGridAnalyzer.hpp"
+#include <bits/stdc++.h> 
```
Don't include this

jeanlego commented on ODIN_II/SRC/HardSoftLogicMixer.cpp:
```diff
@@ -0,0 +1,256 @@
+#include "HardSoftLogicMixer.hpp"
+#include <iostream>
+#include <vector>
+#include <cstdint>
```
Use the POSIX one (stdint.h), same for the other ones

jeanlego commented on ODIN_II/SRC/include/HardSoftLogicMixer.hpp:
```diff
@@ -0,0 +1,105 @@
+     * correspond to the architecture, keyed by the pointers to 
+     * layout  description
+     */
+    std::map<int,std::pair<int,int>> _grid_layout_sizes; 
```
I would prefer a class or a struct over a pair.

jeanlego commented on ODIN_II/SRC/OdinGridAnalyzer.cpp:
```diff
@@ -0,0 +1,334 @@
+	std::cout<< "The type tag is"<<typeTag<<std::endl;
+	std::vector<t_physical_tile_type>::iterator pointer;
+
+	for (pointer =  tileTypes.begin();pointer< tileTypes.end(); pointer++){
```
Use a foreach loop, the iterator one is only used if you are to delete anything from the vector

jeanlego commented on ODIN_II/SRC/OdinGridAnalyzer.cpp:
```diff
@@ -0,0 +1,334 @@
+		//Load the x specification
+		auto& xspec = grid_loc_def->x;
+
+		VTR_ASSERT_MSG(!xspec.start_expr.empty(), "x start position must be specified");
```
Use the Odin one since the VTR assert as compile levels and might be skipped ( I don't know how they do their assertion levels and I'd rather avoid adding the complexity here unless we drop our own version and change over completely to theirs) keeping both is not something I would do, use the oassert or if guard an error_message

georgkrylov committed:
New commit for hard and soft logic mixing

This commit has all required changes to enable working on the
optimization, however there is no history and unnecessary files on it

georgkrylov commented:
This is irrelevant after code updates, but should be kept for reference (please).
